README.md
# xrandr-extend
Extend a HIDPI screen to a normal DPI external display. This command line tool
implements various solutions described in the [HIDPI Arch Linux wiki
page](https://wiki.archlinux.org/index.php/HiDPI#Multiple_displays).


## Installation

```sh
pip install xrandr-extend --user
```
or alternatively

```sh
pip install -e "git+https://github.com/ashwinvis/xrandr-extend.git#egg=xrandr_extend" --user
```

## Configuration
```sh
python -m xrandr_extend.config

This creates a file `~/.config/xrandr-extend.cfg` which looks like this:

```ini
[provider:modesetting]
primary = eDP-1
hdmi = HDMI-1
vga = DP-1

[provider:intel]
primary = eDP1
hdmi = HDMI1
vga = DP1

[resolutions]
primary = (3200, 1800)
hdmi = (1920, 1080)
vga = (1920, 1200)

```

The first few sections have the name in the format `[provider:display_driver]`.
Run `xrandr --listproviders` to find what your system has. The values in this
section should be given as `alias = monitor_name`, as in the output of
`xrandr --listmonitors` command. You may even remove existing sections and 
add more sections for your *display driver*.

Each line in the `[resolutions]` section signifies a *resolution profile* in
the format `alias = [width_in_pixels, height_in_pixels]`.  The *profile*
`primary` should contain the resolution of your built-in display.  You may edit
or remove the remaining values `hdmi` and `vga`.

## Quick reference

```console
usage: xrandr-extend [-h] [-p PRI_RES PRI_RES] [-e EXT_RES EXT_RES] [-m] [-n]
                     [-o] [-s] [-d]
                     profile

Extend a HIDPI screen to a normal DPI external display

positional arguments:
  profile               Use preset external resolution profiles (available:
                        ['hdmi', 'vga']).

optional arguments:
  -h, --help            show this help message and exit
  -p PRI_RES PRI_RES, --pri-res PRI_RES PRI_RES
                        Modify preset resolution of primary display (default
                        [3200, 1800])
  -e EXT_RES EXT_RES, --ext-res EXT_RES EXT_RES
                        Modify preset resolution of ext. display (default
                        based on profile)
  -m, --mirror          Mirror the ext. display
  -n, --pan             pan the position of ext. display
  -o, --only            extend and use only ext. display
  -s, --pos             set the position of ext. display explicitly
  -d, --dry-run         Preview command without executing it

Examples
--------
# Default option requires only the scaling factors and display name
$ xrandr-extend --dry-run vga
$ xrandr-extend vga
$ xrandr-extend hdmi

# Other options to extend the display
$ xrandr-extend --pan hdmi
$ xrandr-extend --only hdmi
$ xrandr-extend -e 1024 768 -n vga  # Pan with custom external resolution
```


CHANGES.md
# Changelog

## 0.0.3

* Deploy to PyPI
* Reorganize as a package and allow for configuration
* Use `pkg_resources` to find `default.cfg`

## 0.0.2

* Simpler defaults which uses only scaling factors
* Parse args only inside `__main__` and do not run any commands during dry run
* Less bugs


xrandr_extend/main.py
#!/usr/bin/env python3
"""
Examples
--------
# Default option requires only the scaling factors and display name
$ xrandr-extend --dry-run vga
$ xrandr-extend vga
$ xrandr-extend hdmi

# Other options to extend the display
$ xrandr-extend --pan hdmi
$ xrandr-extend --only hdmi
$ xrandr-extend -e 1024 768 -n vga  # Pan with custom external resolution

"""
from datetime import datetime
import argparse
from ast import literal_eval
from . import cmd, config


COPYING = """
xrandr-extend Copyright (C) 2018-{} Ashwin Vishnu Mohanan
This program comes with ABSOLUTELY NO WARRANTY; This is free software, and you
are welcome to redistribute it under certain conditions. You should have
received a copy of the GNU General Public License (version 3 or later) along
with this program. If not, see <http://www.gnu.org/licenses/>.
""".format(datetime.now().year)


display_res_defaults = config.read()["resolutions"]


# Parse command-line arguments
parser = argparse.ArgumentParser(
    description="Extend a HIDPI screen to a normal DPI external display",
    epilog=__doc__ + COPYING,
    formatter_class=argparse.RawDescriptionHelpFormatter,
)
parser.add_argument(
    "profile",
    help="Use preset external resolution profiles (available: {}).".format(
        [k for k in display_res_defaults.keys() if k != "primary"]
    ),
)
parser.add_argument(
    "-p",
    "--pri-res",
    help="Modify preset resolution of primary display (default {})".format(
        display_res_defaults["primary"]
    ),
    nargs=2,
    type=int,
    default=literal_eval(display_res_defaults["primary"]),
)
parser.add_argument(
    "-e",
    "--ext-res",
    help="Modify preset resolution of ext. display (default based on profile)",
    nargs=2,
    type=int,
    default=None,
)
parser.add_argument(
    "-m", "--mirror", help="Mirror the ext. display", action="store_true"
)
parser.add_argument(
    "-n", "--pan", help="pan the position of ext. display", action="store_true"
)
parser.add_argument(
    "-o", "--only", help="extend and use only ext. display", action="store_true"
)
parser.add_argument(
    "-s",
    "--pos",
    help="set the position of ext. display explicitly",
    action="store_true",
)
parser.add_argument(
    "-d",
    "--dry-run",
    help="Preview command without executing it",
    action="store_true",
)


def run():
    args = parser.parse_args()

    display_names = cmd.display_names_from_providers()

    monitor1 = display_names["primary"]
    monitor2 = display_names[args.profile]

    # HIDPI monitor resolution
    A = args.pri_res[0]
    B = args.pri_res[1]

    # External non-HIDPI monitor resolution
    if args.ext_res is None:
        ext_res = display_res_defaults[args.profile]
        args.ext_res = literal_eval(ext_res)

    C = args.ext_res[0]
    D = args.ext_res[1]

    # Scaling factor
    E = round(A / C, 2)
    F = round(B / D, 2)
    E = F = max(E, F)

    # Prepare commands
    commands = ["xrandr --auto"]
    commands.extend(["xrandr --listmonitors"])
    if args.mirror:
        commands.extend(
            ["xrandr --output {} --scale {}x{}".format(monitor2, E, F)]
        )
    elif args.pan:
        commands.extend(
            [
                (
                    "xrandr --output {} --auto --output {} --auto --panning "
                    "{}x{}+{}+0 --scale {}x{} --right-of {}"
                ).format(
                    monitor1, monitor2, int(C * E), int(D * F), A, E, F, monitor1
                )
            ]
        )
    elif args.pos:
        commands.extend(
            [
                (
                    "xrandr --output {} --auto --pos 0x{}  --output {} "
                    "--scale {}x{} --auto --pos 0x0 --fb {}x{}"
                ).format(
                    monitor1,
                    int(D * F),
                    monitor2,
                    E,
                    F,
                    int(max(A, C * E)),
                    int(B + D * F),
                )
            ]
        )
    elif args.only:
        commands.extend(["xrandr --output {} --off".format(monitor1)])
    else:
        commands.extend(
            [
                (
                    "xrandr --output {} --auto --output {} --auto --scale {}x{} "
                    "--right-of {}"
                ).format(monitor1, monitor2, E, F, monitor1)
            ]
        )

    list(map(print, commands))
    if not args.dry_run:
        list(map(cmd.call, commands))


if __name__ == "__main__":
    run()
